### PR TITLE
wasm: Add data patching API's to wasm helper bin

### PR DESCRIPTION
**Pull request by patrick-east**
There are three new API's implemented and exposed from the C code:

```
opa_value_add_path
opa_value_remove_path
```

and a new helper:

```
opa_object_remove
```

The first two provide similar functionality as the OPA stores "add"
and "remove" op (same style of path). The main difference for the add
is that it will create intermediate objects as required, we do not
have a separate `mkdir` operation like the store does.

Signed-off-by: Patrick East <east.patrick@gmail.com>

### CODE REVIEW

**Review by tsandall — APPROVED** [review]
LGTM. Just a minor comments on the docs and what looks like a small leak in an edge case.

File: wasm/src/value.c [hunk 1]
```diff
@@ -1403,3 +1432,121 @@ opa_set_elem_t *opa_set_get(opa_set_t *set, opa_value *v)
...
+        {
+            next = opa_object();
+
+            switch (curr->type)
+            {
+                case OPA_OBJECT:
+                    opa_object_insert(opa_cast_object(curr), k, next);
+                    break;
+                default:
+                    return NULL;
```
[the remark this answers is not in the document]
Comment (patrick-east) [reply wasm/src/value.c:1480]: Good catch!

[the remark this answers is not in the document]
Comment (patrick-east) [reply wasm/src/value.c]: Done

File: docs/content/wasm.md [hunk 2]
```diff
@@ -121,6 +121,9 @@ The primary exported functions for interacting with policy modules are:
 | <span class="opa-keep-it-together">`str_addr opa_json_dump(value_addr)`</span> | Dumps the value referred to by `value_addr` to a null-terminated JSON serialized string and returns the address of the start of the string. |
 | <span class="opa-keep-it-together">`void opa_heap_ptr_set(addr)`</span> | Set the heap pointer for the next evaluation. |
 | <span class="opa-keep-it-together">`addr opa_heap_ptr_get(void)`</span> | Get the current heap pointer. |
+| <span class="opa-keep-it-together">`addr opa_value_add_path(base_value_addr, path_value_addr, value_addr)`</span> | Add the value at the `value_addr` into the object referenced by `base_value_addr` at the given path. The `path_value_addr` must point to an array value with string keys (eg: `["a", "b", "c"]`). Existing values will be updated. For example, base object `{"a": {"b": 123}}`, path `["a", "x", "y"]`, and value `{"foo": "bar"}` will yield `{"a": {"b": 123, "x": {"y": {"foo": "bar"}}}}`. Returns 0 on error. |
```
[the remark this answers is not in the document]
Comment (patrick-east) [reply docs/content/wasm.md]: The latest updates include using an error code as well as the doc changes for both error codes and memory management of the parameters. I wasn't 100% how we wanted to do the error codes, definitely open to suggestions.

**Review by tsandall — DISMISSED** [review]
LGTM

**Pull request merged**